athena-google-bigquery/src/main/java/com/amazonaws/athena/connectors/google/bigquery/BigQueryUtils.java

```diff
@@ -233,4 +245,50 @@ static ArrowType translateToArrowType(LegacySQLTypeName type)
                 return new ArrowType.Utf8();
         }
     }
+
+    public static List<Field> getChildFieldList(com.google.cloud.bigquery.Field field)
+    {
+        List<Field> fieldList = new ArrayList<>();
+        if (null != field.getSubFields()) {
+            for (com.google.cloud.bigquery.Field subField : field.getSubFields()) {
+                if (null != subField.getMode() && subField.getMode().name().equals("REPEATED")) {
+                    fieldList.add(new Field(subField.getName(), FieldType.nullable(Types.MinorType.LIST.getType()), getChildFieldList(subField)));
+                }
+                else if (subField.getType().getStandardType().name().equalsIgnoreCase("Struct")) {
+                    fieldList.add(new Field(subField.getName(), FieldType.nullable(Types.MinorType.STRUCT.getType()), getChildFieldList(subField)));
+                }
+                else {
+                    fieldList.add(Field.nullable(subField.getName(), translateToArrowType(subField.getType())));
+                }
+            }
+        }
+        else {
+             fieldList.add(Field.nullable(field.getName(), translateToArrowType(field.getType())));
+        }
+       return  fieldList;
+    }
+
+    public static Object getComplexObjectFromFieldValue(Field field, FieldValue fieldValue, boolean isTimeStampCol) throws ParseException
+    {
+        Types.MinorType minorTypeForArrowType = getMinorTypeForArrowType(field.getFieldType().getType());
+        if (minorTypeForArrowType.equals(Types.MinorType.LIST)) {
+            List<Object> valList = new ArrayList();
+            for (FieldValue fieldVal : fieldValue.getRepeatedValue()) {
+                valList.add(getObjectFromFieldValue(field.toString(), fieldVal, field.getChildren().get(0), isTimeStampCol));
```

Comment:
I may not be following. Can it be clarified how `getComplexObjectFromFieldValue` and `getObjectFromFieldValue` differ? It appears that both functions call each other, resulting in indirect recursion. Is there a reason why it can't be combined into one recursive function?

Reply:
yes, it is indirect recursion, only to read child values from complex datatypes, like a list of structs, any level of complexity in values. 
It is getting called from child datatypes from each parent. @akuzin1